# Unveiling the Driving Factors: Exploring the Influences on Car Prices

As an analyst at Crankshaft List, delving into the world of car sales is an exhilarating experience. Each day, our platform witnesses a surge of free vehicle advertisements, painting a vivid picture of the market's dynamics.Through a comprehensive analysis of our vast vehicles ads database, we aim to shed light on the pivotal factors that wield the greatest influence when it comes to selling a car.

### Table of content

<span style='color:blue'>

1.Introduction
  

2.General Information


3.First Impressions
 
4.Data Preprocessing
  
5.Exploratory Data Analysis

6.Unveiling Top Car Types
  
 7.Unraveling Price Factors
  
8.General Conclusions: Insights and Takeaways

9.Conclusion and Project Completion Checklist

### Introduction

Welcome to the fascinating world of car sales analysis! In this analysis, we will delve into car sales data and embark on a journey to uncover the factors that play a pivotal role in influencing car prices.

Car sales are a dynamic and ever-evolving market, with hundreds of transactions taking place every day. Understanding the intricate web of variables that contribute to price fluctuations is crucial for both buyers and sellers in making informed decisions.

### General information

In [182]:
# Loading all the libraries
import pandas as pd
import plotly.express as px
from pandas.plotting import scatter_matrix
import datetime
import numpy as np




Load the dataset.

In [147]:
try:
    vehicle_ad_data=pd.read_csv('/Users/lelou/Car-Advertisement/vehicles_us.csv')
except:
    vehicle_ad_data=pd.read_csv('https://code.s3.yandex.net/datasets/vehicles_us.csv')

In [148]:
vehicle_ad_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB



The "vehicle_us" dataset consists of 51,525 entries with 13 columns, providing information about vehicles available for sale. It includes details like price, model year, model, condition, cylinders, fuel, odometer, transmission, type, paint color, is_4wd, date posted, and days listed. The dataset contains a mix of numeric and categorical data, with some columns having missing values. Overall, analyzing this dataset can offer insights into the used vehicle market and factors affecting pricing and desirability.

### First Impressions

In [149]:
#look at a sample of data for first 20 rows
vehicle_ad_data.head(20)

price  model_year                     model  condition  cylinders fuel  \
0    9400      2011.0                    bmw x5       good        6.0  gas   
1   25500         NaN                ford f-150       good        6.0  gas   
2    5500      2013.0            hyundai sonata   like new        4.0  gas   
3    1500      2003.0                ford f-150       fair        8.0  gas   
4   14900      2017.0              chrysler 200  excellent        4.0  gas   
5   14990      2014.0              chrysler 300  excellent        6.0  gas   
6   12990      2015.0              toyota camry  excellent        4.0  gas   
7   15990      2013.0               honda pilot  excellent        6.0  gas   
8   11500      2012.0               kia sorento  excellent        4.0  gas   
9    9200      2008.0               honda pilot  excellent        NaN  gas   
10  19500      2011.0  chevrolet silverado 1500  excellent        8.0  gas   
11   8990      2012.0              honda accord  excellent        4.0  gas   
12  18990      2012.0                  ram 1500  excellent        8.0  gas   
13  16500      2018.0            hyundai sonata  excellent        4.0  gas   
14  12990      2009.0                 gmc yukon  excellent        8.0  gas   
15  17990      2013.0                  ram 1500  excellent        8.0  gas   
16  14990      2010.0                  ram 1500  excellent        8.0  gas   
17  13990      2014.0             jeep cherokee  excellent        6.0  gas   
18  12500      2013.0        chevrolet traverse  excellent        6.0  gas   
19  13990      2018.0           hyundai elantra  excellent        4.0  gas   

    odometer transmission    type paint_color  is_4wd date_posted  days_listed  
0   145000.0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    88705.0    automatic  pickup       white     1.0  2018-10-19           50  
2   110000.0    automatic   sedan         red     NaN  2019-02-07           79  
3        NaN    automatic  pickup         NaN     NaN  2019-03-22            9  
4    80903.0    automatic   sedan       black     NaN  2019-04-02           28  
5    57954.0    automatic   sedan       black     1.0  2018-06-20           15  
6    79212.0    automatic   sedan       white     NaN  2018-12-27           73  
7   109473.0    automatic     SUV       black     1.0  2019-01-07           68  
8   104174.0    automatic     SUV         NaN     1.0  2018-07-16           19  
9   147191.0    automatic     SUV        blue     1.0  2019-02-15           17  
10  128413.0    automatic  pickup       black     1.0  2018-09-17           38  
11  111142.0    automatic   sedan        grey     NaN  2019-03-28           29  
12  140742.0    automatic  pickup         NaN     1.0  2019-04-02           37  
13   22104.0    automatic   sedan      silver     NaN  2019-01-14           29  
14  132285.0    automatic     SUV       black     1.0  2019-01-31           24  
15       NaN    automatic  pickup         red     1.0  2018-05-15          111  
16  130725.0    automatic  pickup         red     1.0  2018-12-30           13  
17  100669.0    automatic     SUV         red     1.0  2018-08-16           25  
18  128325.0    automatic     SUV       white     1.0  2019-04-09           13  
19   31932.0    automatic   sedan         red     NaN  2018-08-25           27

In [150]:
# A summary statistics table for categorical columns in a DataFrame
vehicle_ad_data.describe(include='object').T

count unique         top   freq
model         51525    100  ford f-150   2796
condition     51525      6   excellent  24773
fuel          51525      5         gas  47288
transmission  51525      3   automatic  46902
type          51525     13         SUV  12405
paint_color   42258     12       white  10029
date_posted   51525    354  2019-03-17    186

The dataset contains information on 51,525 vehicles. The most popular vehicle model is the "Ford F-150" with 2,796 occurrences. A significant number of vehicles are listed as being in "excellent" condition (24,773 entries). The majority of vehicles run on "gas" fuel (47,288 entries) and have "automatic" transmission (46,902 entries). The most frequently listed vehicle type is "SUV" (12,405 entries). The most common paint color is "white" (10,029 entries), and the vehicles were posted on 354 different dates, with the most frequent date being "2019-03-17" (186 occurrences).

## Data Preprocessing

### Processing missing value

In [151]:
#count number of null in the dataset
print('total number of row :',len(vehicle_ad_data))
vehicle_ad_data.isna().sum()


total number of row : 51525


price               0
model_year       3619
model               0
condition           0
cylinders        5260
fuel                0
odometer         7892
transmission        0
type                0
paint_color      9267
is_4wd          25953
date_posted         0
days_listed         0
dtype: int64

In [152]:
#Propotion of missing value compare to the whole percentage in percent
propor_miss_percent = (vehicle_ad_data.isnull().sum() / len(vehicle_ad_data)) * 100
propor_miss_percent

price            0.000000
model_year       7.023775
model            0.000000
condition        0.000000
cylinders       10.208637
fuel             0.000000
odometer        15.316836
transmission     0.000000
type             0.000000
paint_color     17.985444
is_4wd          50.369723
date_posted      0.000000
days_listed      0.000000
dtype: float64

<span style='color:blue'>
The dataset analysis reveals intriguing insights:

-Missing values are present across multiple columns, except for the "price" column which is complete.

-The "model_year" column has 3,619 missing values, creating uncertainty about the manufacturing year of certain vehicles.

-The "cylinders" column shows 5,260 missing values, suggesting a lack of cylinder information.

-The "odometer" column has 7,892 missing values, indicating a lack of distance traveled data.

-The "paint_color" column has 9,267 missing values, leaving color information unknown.

-Surprisingly, the "is_4wd" column has 25,953 missing values, with almost half of the available values being 1. These missing values could potentially be interpreted as 0.

-Possible reasons for missing values include intentional omission, data entry errors, incomplete information from sellers, and even incorrect measurements or equipment errors.

-Anomalies like unusually low prices raise concerns about data quality.
Effective data cleaning and imputation techniques are crucial to ensure accurate and comprehensive analysis.

### Step to fill in missing data

In [153]:
# Fill the NaNs in is_4wd column with zeros since the column contains either 1 or 0
vehicle_ad_data['is_4wd'].fillna(value= 0.0, inplace=True)
vehicle_ad_data['is_4wd'] = vehicle_ad_data['is_4wd'].astype(int)
vehicle_ad_data.head()

price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN       1  2018-06-23           19  
1    automatic  pickup       white       1  2018-10-19           50  
2    automatic   sedan         red       0  2019-02-07           79  
3    automatic  pickup         NaN       0  2019-03-22            9  
4    automatic   sedan       black       0  2019-04-02           28

In [154]:
#Fill missing value in model_year column with the median of cars of the same model
vehicle_ad_data['model_year'].fillna(
    vehicle_ad_data.groupby(['model'])['model_year'].transform(np.median), inplace=True
)

By using the median to fill in missing 'model_year' values, we are choosing a value that is representative of the central tendency of the data. It provides a reasonable estimate for the missing values without being skewed by extreme values which is 'model_year' min value is 1908 and max value 2019 

In [155]:
#Fill missing values in cylinders based on car model and model year
from statistics import mean
vehicle_ad_data['cylinders'].fillna(
    vehicle_ad_data.groupby(['model'])['model_year'].transform(mean), inplace=True
)


The cylinders column represents discrete values, and using the mean could be appropriate as there no outliers or extreme values that might heavily influence the mean as both mean value of 'cylinders' is 6 which is in reasonable range compare to the min value 3 and max value 12

In [156]:
#Filled last missing odometer values of same listing with the mean value 
vehicle_ad_data['odometer'].fillna((vehicle_ad_data['odometer'].mean()), inplace=True)

In [157]:
# change the missing values from 'paint_color' with the most occuring color from mode
vehicle_ad_data['paint_color'] = vehicle_ad_data.groupby('model')['paint_color'].transform(lambda x: x.fillna(x.mode().iloc[0]))

In [158]:
vehicle_ad_data.describe()

price    model_year     cylinders       odometer        is_4wd  \
count   51525.000000  51525.000000  51525.000000   51525.000000  51525.000000   
mean    12132.464920   2009.793954    210.673715  115553.461738      0.496303   
std     10040.803015      6.099296    606.646534   59902.205520      0.499991   
min         1.000000   1908.000000      3.000000       0.000000      0.000000   
25%      5000.000000   2007.000000      4.000000   79181.000000      0.000000   
50%      9000.000000   2011.000000      6.000000  115553.461738      0.000000   
75%     16839.000000   2014.000000      8.000000  146541.000000      1.000000   
max    375000.000000   2019.000000   2017.463768  990000.000000      1.000000   

       days_listed  
count  51525.00000  
mean      39.55476  
std       28.20427  
min        0.00000  
25%       19.00000  
50%       33.00000  
75%       53.00000  
max      271.00000

In [159]:
# Make sure the table contains no more missing columns
missing_counts = vehicle_ad_data.isna().sum()
# Display the count of missing values in each column
print( missing_counts)

price           0
model_year      0
model           0
condition       0
cylinders       0
fuel            0
odometer        0
transmission    0
type            0
paint_color     0
is_4wd          0
date_posted     0
days_listed     0
dtype: int64


### Process Duplicates

In [160]:
# counting clear dublicates
print(vehicle_ad_data.duplicated().sum()) 

0


The result indicates that there are no obvious duplicates present in the dataset. However, it is possible that there are implicit duplicates where the model names are written differently. Such errors can impact the accuracy of the results obtained from the analysis.

In [176]:
#viewing unique model name
vehicle_ad_data['model'].unique()

array(['bmw x5', 'ford f-150', 'hyundai sonata', 'chrysler 200',
       'chrysler 300', 'toyota camry', 'honda pilot', 'kia sorento',
       'chevrolet silverado 1500', 'honda accord', 'ram 1500',
       'gmc yukon', 'jeep cherokee', 'chevrolet traverse',
       'hyundai elantra', 'chevrolet tahoe', 'toyota rav4',
       'chevrolet silverado', 'jeep wrangler', 'chevrolet malibu',
       'ford fusion se', 'chevrolet impala', 'chevrolet corvette',
       'jeep liberty', 'toyota camry le', 'nissan altima',
       'subaru outback', 'toyota highlander', 'dodge charger',
       'toyota tacoma', 'chevrolet equinox', 'nissan rogue',
       'mercedes-benz benze sprinter 2500', 'honda cr-v',
       'jeep grand cherokee', 'toyota 4runner', 'ford focus',
       'honda civic', 'kia soul', 'chevrolet colorado',
       'ford f150 supercrew cab xlt', 'chevrolet camaro lt coupe 2d',
       'chevrolet cruze', 'ford mustang', 'chevrolet silverado 3500hd',
       'nissan frontier crew cab sv', 'subaru imp

Based on the 'model' column, we found duplicates for the Ford F-150, Ford F-250 SD, and Ford F-350 SD. The model names ['ford f-150', 'ford f-250 sd', 'ford f-350 sd'] correspond to ['ford f150', 'ford f-250 super duty', 'ford f350 super duty'].
To get rid of them, declare the function `replace_wrong_models()` with two parameters: 
* `wrong_models=` — the list of duplicates
* `correct_models=` — the string with the correct value

The function should correct the names in the `'model'` column from the `vehicle_ad_data` table.


In [ ]:
# function replace implicit duplicates
def replace_wrong_models(wrong_models, correct_models):
    # Replace the wrong model names with the correct model names
    vehicle_ad_data['model'] = vehicle_ad_data['model'].replace(wrong_models, correct_models)
    # Print the updated 'model' column
    print(vehicle_ad_data['model'])


In [ ]:
#list of wrong model
wrong_models=['ford f150', 'ford f-250 super duty', 'ford f350 super duty']
#list of correct model
correct_models=['ford f-150', 'ford f-250 sd', 'ford f-350 sd']
#call of the function to replace wrong models
replace_wrong_models(wrong_models, correct_models)

0                bmw x5
1            ford f-150
2        hyundai sonata
3            ford f-150
4          chrysler 200
              ...      
51520     nissan maxima
51521       honda civic
51522    hyundai sonata
51523    toyota corolla
51524     nissan altima
Name: model, Length: 40840, dtype: object


Last but not least, we prefer to have a unique combinations of 'price', 'model_year', 'model', and 'odometer' values which will eliminate  duplicate data. As duplicate data can be a sign of data quality issues, such as data entry errors or data duplication during merging or importing processes. Identifying and removing duplicates improves the overall quality of the dataset.

In [ ]:
#Drop duplicates
vehicle_ad_data = vehicle_ad_data.drop_duplicates(subset=['model','price', 'model_year', 'model', 'odometer'])
vehicle_ad_data.info()


<class 'pandas.core.frame.DataFrame'>
Index: 40699 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         40699 non-null  int64  
 1   model_year    40699 non-null  float64
 2   model         40699 non-null  object 
 3   condition     40699 non-null  object 
 4   cylinders     40699 non-null  float64
 5   fuel          40699 non-null  object 
 6   odometer      40699 non-null  float64
 7   transmission  40699 non-null  object 
 8   type          40699 non-null  object 
 9   paint_color   40699 non-null  object 
 10  is_4wd        40699 non-null  int32  
 11  date_posted   40699 non-null  object 
 12  days_listed   40699 non-null  int64  
dtypes: float64(3), int32(1), int64(2), object(7)
memory usage: 4.2+ MB


### Data replacement

In [180]:
#check data type
vehicle_ad_data.dtypes

price                    int64
model_year               int32
model                   object
condition               object
cylinders                int32
fuel                    object
odometer                 int32
transmission            object
type                    object
paint_color             object
is_4wd                   int32
date_posted     datetime64[ns]
days_listed              int64
year_posted              int32
Car_Age                  int64
Car_mileage              Int64
dtype: object

model_year: The model_year column is currently represented as a float64 data type. To plot a histogram or scatterplot based on the model year, it is generally more appropriate to convert it to an integer data type. This allows for better visualization and interpretation of the data, as model years are typically represented as whole numbers.

cylinders: The cylinders column is currently represented as a float64 data type. However, since the number of cylinders in a vehicle is typically a discrete value, it is often more appropriate to convert it to an integer data type for plotting purposes. This ensures that the data is displayed accurately without decimal places.

odometer: The odometer column is currently represented as a float64 data type, which indicates a continuous numerical value. To plot a histogram or scatterplot based on the odometer reading, it is common to keep it as a numerical value rather than converting it to a different data type.

date_posted: The date_posted column is currently represented as an object data type, which typically indicates a string or generic object. To plot data over time or analyze temporal patterns, it is advisable to convert this column to a datetime data type. This allows for easier manipulation, sorting, and visualization of the dates.

In [174]:
#Replaceing data type for model_year category to int
vehicle_ad_data['model_year'] = vehicle_ad_data['model_year'].astype(int)
#Replaceing data type for cylinders category to int
vehicle_ad_data['cylinders'] = vehicle_ad_data['cylinders'].astype(int)
#Replaceing data type for odometer category to int
vehicle_ad_data['odometer'] = vehicle_ad_data['odometer'].astype(int)
#Replacing data type for date_posted into datetime type
vehicle_ad_data['date_posted']=pd.to_datetime(vehicle_ad_data['date_posted'])

In [175]:
# Check for data type again
#Checking data types
vehicle_ad_data.dtypes

price                    int64
model_year               int32
model                   object
condition               object
cylinders                int32
fuel                    object
odometer                 int32
transmission            object
type                    object
paint_color             object
is_4wd                   int32
date_posted     datetime64[ns]
days_listed              int64
dtype: object

### Make Calculation and Enrich the data

In [ ]:
#  change the condition values with something that can be manipulated more easily
condition_mapping = {'new': 5, 'like new': 4, 'excellent': 3, 'good': 2, 'fair': 1, 'salvage': 0}
vehicle_ad_data['condition'] = vehicle_ad_data['condition'].replace(condition_mapping)


In [187]:
#add the age of the car
# first isolate 'year' from 'date_posted' column
vehicle_ad_data['year_posted'] = vehicle_ad_data['date_posted'].dt.year
# create a function that calculate car age
def calculate_car_age(row):
    return (row['year_posted'] - row['model_year'])+1

vehicle_ad_data['Car_Age'] = vehicle_ad_data.apply(calculate_car_age, axis=1)

vehicle_ad_data['Car_Age'].unique()


array([  8,   7,  17,   3,   5,   4,  12,   2,  11,   6,   9,   1,  10,
        16,  15,  14,  18,  13,  54,  25,  19,  20,  27,  23,  39,  21,
        24,  45,  22,  26,  34,  43,  32,  44,  30,  29,  47,  53,  50,
        31,  41,  55,  40,  51,  28,  33,  56,  38,  57,  35,  46,  42,
        52,  64,  48,  49,  37,  36,  66,  58,  72,  59,  60, 111,  83,
        70,  61,  90], dtype=int64)

Mileage is a crucial factor that can significantly impact the price of a car. Generally, higher mileage tends to lower the price. To calculate the car's mileage, we utilize the following function:

vehicle_data['Car_mileage'] = vehicle_data['odometer'] / (vehicle_data['year_posted'] - vehicle_data['model_year'])

In this function, we include a "+1" to the equation row['year_posted'] - row['model_year']. This addition ensures that we avoid infinite values for 'Car_mileage'. Without this adjustment, if there are 'Car_age' values of 0, the equation would result in division by zero, which is undefined.

In [186]:
# Add 'Car_mileage' and handle non-finite values
vehicle_ad_data['Car_mileage'] = vehicle_ad_data['odometer'] / (vehicle_ad_data['year_posted'] - vehicle_ad_data['model_year'])
vehicle_ad_data['Car_mileage'].replace([np.inf, -np.inf], np.nan, inplace=True)

# Round 'Car_mileage' to remove decimal places
vehicle_ad_data['Car_mileage'] = vehicle_ad_data['Car_mileage'].round().astype('Int64')

# Check the data type of 'Car_mileage'
vehicle_ad_data['Car_mileage'].dtypes
#Verify new columns
vehicle_ad_data.head()
vehicle_ad_data.info()
st.subheader("Price vs. Odometer Scatter Plot")
fig_scatter = px.scatter(vehicle_ad_data, x='odometer', y='price')
st.plotly_chart(fig_scatter)

# Task 3: Calculate the duration of advertisements
vehicle_ad_data['date_posted'] = pd.to_datetime(vehicle_ad_data['date_posted'])
vehicle_ad_data['days_listed'] = pd.to_timedelta(vehicle_ad_data['days_listed'], unit='D')
vehicle_ad_data['ad_duration'] = vehicle_ad_data['date_posted'] + vehicle_ad_data['days_listed']

# Task 4: Average price across different vehicle types
st.header("Average Price by Vehicle Type")
avg_price_by_type = vehicle_ad_data.groupby('type')['price'].mean().reset_index()
fig_avg_price = px.bar(avg_price_by_type, x='type', y='price', color='type')
st.plotly_chart(fig_avg_price)

# Task 5: Checkbox to toggle display
display_details = st.checkbox("Display Data Details")
if display_details:
    st.subheader("Data Details")
    st.write(vehicle_ad_data.head())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 16 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   price         51525 non-null  int64         
 1   model_year    51525 non-null  int32         
 2   model         51525 non-null  object        
 3   condition     51525 non-null  object        
 4   cylinders     51525 non-null  int32         
 5   fuel          51525 non-null  object        
 6   odometer      51525 non-null  int32         
 7   transmission  51525 non-null  object        
 8   type          51525 non-null  object        
 9   paint_color   51525 non-null  object        
 10  is_4wd        51525 non-null  int32         
 11  date_posted   51525 non-null  datetime64[ns]
 12  days_listed   51525 non-null  int64         
 13  year_posted   51525 non-null  int32         
 14  Car_Age       51525 non-null  int64         
 15  Car_mileage   49373 non-null  Int64 

### Exploratory data analysis

In [178]:
# Task 1: Investigate outliers in numerical columns using box plots
st.header("Outliers Investigation")
st.subheader("Price Box Plot")
fig_price = px.box(vehicle_ad_data, y='price')
print(st.plotly_chart(fig_price))


DeltaGenerator()


In [185]:
# Task 2: Analyze core parameters using histograms and scatter plots
st.header("Core Parameters Analysis")
st.subheader("Price Distribution")
fig_price_hist = px.histogram(vehicle_ad_data, x='price')
st.plotly_chart(fig_price_hist)

st.subheader("Odometer Distribution")
fig_odometer_hist = px.histogram(vehicle_ad_data, x='odometer')
st.plotly_chart(fig_odometer_hist)

st.subheader("Price vs. Odometer Scatter Plot")
fig_scatter = px.scatter(vehicle_ad_data, x='odometer', y='price')
st.plotly_chart(fig_scatter)

DeltaGenerator()